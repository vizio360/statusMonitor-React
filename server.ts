let express = require('express');
let httpProxy = require('http-proxy');
let path = require('path');
let fs = require('fs');
const url = require('url');
var bodyParser = require('body-parser');

let apiProxy = httpProxy.createProxyServer();

let app = express();

app.use(bodyParser.json());
app.use(express.static(path.join(__dirname, 'dist')));

app.get('/forward', (req: any, res: any) => {
  let targetUrl: URL = url.parse(req.query.uri);
  console.log('forwarding ' + targetUrl.pathname);
  req.url = targetUrl.pathname;

  let target: string = targetUrl.protocol + '//' + targetUrl.host;

  apiProxy.web(req, res, {target: target}, function(e: any) {
    console.log('proxying went wrong');
    console.log(e);
    res.sendStatus(500);
  });
});

let getFileContent = function(def: string, path: string) {
  if (fs.existsSync(path)) {
    def = fs.readFileSync(path, {encoding: 'utf8'});
  }
  return JSON.parse(def);
};

app.get('/services', (req: any, res: any) => {
  let path: string = './config/services.json';
  let returnData = getFileContent('{"services":[]}', path);
  res.json(returnData.services);
});
app.get('/connections', (req: any, res: any) => {
  let path: string = './config/connections.json';
  let returnData = getFileContent('{"connections":[]}', path);
  res.json(returnData.connections);
});
app.post('/services', (req: any, res: any) => {
  console.log(req.body);
  let data = req.body;
  console.log(data);
  fs.writeFileSync('./config/services.json', JSON.stringify(data), {
    flag: 'w+',
  });
  res.sendStatus(200);
});

app.post('/connections', (req: any, res: any) => {
  console.log(req.body);
  let data = req.body;
  console.log(data);
  fs.writeFileSync('./config/connections.json', JSON.stringify(data), {
    flag: 'w+',
  });
  res.sendStatus(200);
});

//read config services and connections
//store config in data structure
//set up timers for services
//on service status changed broadcast message to clients
//

let port: number = 3333;
app.listen(port);
console.log('Listening to ' + port);
