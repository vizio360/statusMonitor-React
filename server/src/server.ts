import express from 'express';
import http from 'http';
import httpProxy from 'http-proxy';
import path from 'path';
import WebSocket from 'ws';
import {URL} from 'url';
import bodyParser from 'body-parser';
import expressWS from 'express-ws';
import axios from 'axios';
import {
  Command,
  IMessage,
  Reply,
  IService,
  IServiceLastKnownState,
  IConnection,
  Status,
} from '@dataTypes';
import ObjectMatcher from '@server/ObjectMatcher';
import _ from 'underscore';
const enableDestroy = require('server-destroy');

enum ConfigPaths {
  SERVICES = '/services',
  CONNECTIONS = '/connections',
}

class StatusMonitoringServer {
  app: express.Application;
  serverInstance: any;
  appws: expressWS.Application;
  websocketServer: WebSocket.Server;
  services: IService[] = [];
  servicesStatus: IServiceLastKnownState[] = [];
  timeouts: {[serviceId: string]: NodeJS.Timer} = {};
  connections: IConnection[] = [];
  private configApiUrl: string;
  private port: number;
  private proxyServer: httpProxy;

  constructor() {
    this.app = express();
    const expWS = expressWS(this.app);
    this.websocketServer = expWS.getWss();
    this.appws = expWS.app;
    this.app.use(express.static(path.join(__dirname, '../../dist')));
  }

  init(listeningPort: number) {
    this.listen(listeningPort);
    this.setupWebSockets();
    this.setupRestAPI();
  }

  private setupProxy(target: string) {
    this.proxyServer = httpProxy.createProxyServer();
    this.app.post('/config', (req, res) => {
      this.proxyServer.web(req, res, {target: target});
    });
    this.app.use(bodyParser.json());
  }

  private setupRestAPI() {
    this.app.post('/reload', (req, res) => {
      this.start(this.configApiUrl)
        .then(result => {
          const msg: IMessage = {
            reply: Reply.RELOADED,
          };
          this.broadcastMessage(msg);
        })
        .catch(error => {
          const msg: IMessage = {
            reply: Reply.RELOAD_ERROR,
          };
          this.broadcastMessage(msg);
        });
      res.sendStatus(200);
    });
  }

  private listen(listeningPort: number) {
    this.port = listeningPort;
    this.serverInstance = this.app.listen(this.port);
    enableDestroy(this.serverInstance);
  }

  public start(configApiUrl: string): Promise<any> {
    this.configApiUrl = configApiUrl;
    this.setupProxy(configApiUrl);
    this.clearAnyRunningTimeouts();
    return axios
      .all([
        axios.get(this.configApiUrl + ConfigPaths.SERVICES),
        axios.get(this.configApiUrl + ConfigPaths.CONNECTIONS),
      ])
      .then((values: any[]) => {
        this.services = values[0].data;
        this.connections = values[1].data;
        this.setupWatchdogs(this.services);
      });
  }

  public getServicesStatus(): IServiceLastKnownState[] {
    return _.map(this.servicesStatus, _.clone);
  }

  public getServicesStatusById(id: string): IServiceLastKnownState {
    const index: number = _.findIndex(this.servicesStatus, {serviceId: id});
    return Object.assign({}, this.servicesStatus[index]);
  }

  private notifyIfServiceStatusUpdated(
    serviceId: string,
    newStatus: Status,
    responseBody: string,
  ) {
    const ss: IServiceLastKnownState = _.findWhere(this.servicesStatus, {
      serviceId: serviceId,
    });
    if (ss.status != newStatus) {
      const statusReport: IServiceLastKnownState = {
        serviceId: serviceId,
        status: newStatus,
        responseBody: responseBody,
      };
      let msg: IMessage = {
        reply: Reply.UPDATE,
        content: statusReport,
      };
      const index: number = _.findIndex(this.servicesStatus, {
        serviceId: serviceId,
      });
      this.servicesStatus[index] = statusReport;
      this.broadcastMessage(msg);
    }
  }

  private pollHealthcheck(service: IService): void {
    let t: NodeJS.Timer = setTimeout(() => {
      axios
        .get(service.uri)
        .then(response => {
          const responseBody = response.data;
          const state: Status = ObjectMatcher(responseBody, service.matcher)
            ? Status.HEALTHY
            : Status.UNHEALTHY;
          this.notifyIfServiceStatusUpdated(service.id, state, responseBody);
        })
        .catch(error => {
          this.notifyIfServiceStatusUpdated(
            service.id,
            Status.UNHEALTHY,
            error,
          );
        })
        .finally(() => {
          this.pollHealthcheck(service);
        });
    }, service.timeout * 1000);
    this.timeouts[service.id] = t;
  }

  private setupWatchdogs(services: IService[]): void {
    //TODO unit test this reset of servicesStatus
    this.servicesStatus = [];
    services.forEach(service => {
      const statusReport: IServiceLastKnownState = {
        serviceId: service.id,
        status: Status.HEALTHY,
        responseBody: 'NOT YET VERIFIED',
      };
      this.servicesStatus.push(statusReport);
      this.pollHealthcheck(service);
    });
  }

  public broadcastMessage(msg: IMessage): void {
    this.websocketServer.clients.forEach(client => {
      client.send(JSON.stringify(msg));
    });
  }

  private disconnectAllClients(): void {
    this.websocketServer.clients.forEach(client => {
      client.removeAllListeners();
      client.close();
    });
  }

  private setupWebSockets(): void {
    this.appws.ws('/channel', (ws: WebSocket, req: any) => {
      ws.on('message', (command: string) => {
        let msg: IMessage;
        switch (command) {
          case Command.GET_SERVICES:
            msg = {
              reply: Reply.SERVICES,
              content: this.services,
            };
            break;
          case Command.GET_CONNECTIONS:
            msg = {
              reply: Reply.CONNECTIONS,
              content: this.connections,
            };
            break;
          case Command.GET_CURRENT_STATES:
            msg = {
              reply: Reply.CURRENT_STATES,
              content: this.getServicesStatus(),
            };
            break;
          default:
            msg = {
              reply: Reply.NOT_RECOGNIZED,
            };
        }
        ws.send(JSON.stringify(msg));
      });
    });
  }

  private clearAnyRunningTimeouts() {
    for (let key in this.timeouts) {
      clearTimeout(this.timeouts[key]);
    }
    this.timeouts = {};
  }

  public destroy(cb: any): void {
    this.clearAnyRunningTimeouts();
    this.disconnectAllClients();
    if (this.serverInstance) {
      this.serverInstance.destroy(cb);
    }
  }
}

export {StatusMonitoringServer, IMessage, Reply};
