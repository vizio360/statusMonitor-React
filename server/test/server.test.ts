import {StatusMonitoringServer, IMessage, Reply} from '@server/server';
import WebSocket from 'ws';
import * as express from 'express';
import * as DataTypes from '@dataTypes';
import nock from 'nock';
import fs from 'fs';
import path from 'path';
import axios from 'axios';

describe('Status Monitoring Server', () => {
  const CONFIG_URI: string = 'http://configapi';
  const listeningPort: number = 4000;
  const connectionString: string = `ws://localhost:${listeningPort}/channel`;
  const LOCAL_SERVER: string = `http://localhost:${listeningPort}`;

  let getFileContentAsJSON = function(file: string) {
    let content: string = 'File does not exist';
    let fullPath: string = path.resolve(__dirname, file);
    if (fs.existsSync(fullPath)) {
      content = fs.readFileSync(fullPath, {encoding: 'utf8'});
    }
    return JSON.parse(content);
  };

  let setupMockConfigAPI_GET = function(
    resource: string,
    statusCode: number,
    responseBody: object,
  ) {
    nock(CONFIG_URI)
      .get(resource)
      .reply(statusCode, responseBody);
  };

  let setupMockConfigAPI_POST = function(
    resource: string,
    statusCode: number,
    responseBody: object,
  ) {
    nock(CONFIG_URI)
      .post(resource)
      .reply(statusCode, responseBody);
  };

  let server: StatusMonitoringServer = null;
  let activeWebSocketClients: WebSocketClient[] = [];

  beforeEach(() => {
    activeWebSocketClients = [];
    server = new StatusMonitoringServer();
    server.init(listeningPort);
  });

  afterEach(done => {
    activeWebSocketClients.forEach(wsc => {
      wsc.destroy();
    });
    server.destroy(done);
  });

  let setupServicesAndConnectionsMocks = (
    servicesStatus: number = 200,
    connectionsStatus: number = 200,
    servicesFile: string = 'services.json',
    connectionsFile: string = 'connections.json',
  ) => {
    setupMockConfigAPI_GET(
      '/services',
      servicesStatus,
      getFileContentAsJSON(`./mocks/${servicesFile}`),
    );

    setupMockConfigAPI_GET(
      '/connections',
      connectionsStatus,
      getFileContentAsJSON(`./mocks/${connectionsFile}`),
    );
  };

  class WebSocketClient {
    ws: WebSocket;
    onMessage: (data: string) => void;
    public connect(connection: string) {
      this.ws = new WebSocket(connection);
      activeWebSocketClients.push(this);

      return new Promise((resolve, reject) => {
        this.ws.on('open', () => {
          resolve('open');
        });
        this.ws.on('message', (data: string) => {
          if (this.onMessage) this.onMessage(data);
        });
      });
    }

    public send(command: string) {
      this.ws.send(command);
    }

    public destroy() {
      this.ws.removeAllListeners();
    }
  }

  let connectAndSendCommand = (command: string) => {
    return new Promise((resolve, reject) => {
      server
        .start(CONFIG_URI)
        .then(() => {
          const wsc: WebSocketClient = new WebSocketClient();
          wsc.onMessage = (data: string) => {
            resolve(JSON.parse(data));
          };
          wsc.connect(connectionString).then(result => wsc.send(command));
        })
        .catch(error => {
          reject(error);
        });
    });
  };

  test('throws an error if initialised more than once', () => {
    expect(() => server.init(1234)).toThrow();
  });

  test('it can broadcast a message to all connected clients', done => {
    let numberOfCalls: number = 2;
    let callback = (data: string) => {
      numberOfCalls--;
      if (numberOfCalls == 0) {
        done();
      }
    };

    const wsc1: WebSocketClient = new WebSocketClient();
    wsc1.onMessage = callback;

    const wsc2: WebSocketClient = new WebSocketClient();
    wsc2.onMessage = callback;

    Promise.all([
      wsc1.connect(connectionString),
      wsc2.connect(connectionString),
    ]).then(values => {
      let s: DataTypes.IService[] = [];
      let msg = {
        reply: Reply.SERVICES,
        content: s,
      };
      server.broadcastMessage(msg);
    });
  });

  test('sends services config', done => {
    setupServicesAndConnectionsMocks();
    connectAndSendCommand('GET_SERVICES')
      .then((data: IMessage) => {
        expect(data).toMatchSnapshot();
        done();
      })
      .catch(error => {
        fail(error);
        done();
      });
  });

  test('sends connections config', done => {
    setupServicesAndConnectionsMocks();
    connectAndSendCommand('GET_CONNECTIONS')
      .then((data: IMessage) => {
        expect(data).toMatchSnapshot();
        done();
      })
      .catch(error => {
        fail(error);
        done();
      });
  });

  test('start fails if cannot get services data', done => {
    setupServicesAndConnectionsMocks(500);
    connectAndSendCommand('GET_SERVICES').catch(error => {
      done();
    });
  });

  test('start fails if cannot get connections data', done => {
    setupServicesAndConnectionsMocks(200, 500);
    connectAndSendCommand('GET_CONNECTIONS')
      .then((data: IMessage) => {
        fail();
        done();
      })
      .catch(error => {
        done();
      });
  });

  test('handles not existing commands', done => {
    setupServicesAndConnectionsMocks();
    connectAndSendCommand('SOME_UNKNOWN_COMMAND')
      .then((data: IMessage) => {
        expect(data).toMatchSnapshot();
        done();
      })
      .catch(error => {
        fail();
        done();
      });
  });

  test('sends all services status', done => {
    setupServicesAndConnectionsMocks();
    connectAndSendCommand('GET_CURRENT_STATES')
      .then((data: IMessage) => {
        expect(data.reply).toBe('CURRENT_STATES');
        expect(data).toMatchSnapshot();
        done();
      })
      .catch(error => {
        fail(error);
        done();
      });
  });

  let setupServiceHealthcheckMock = (
    url: URL,
    statusCode: number,
    responseBody: any,
  ) => {
    nock(url.origin)
      .persist()
      .get(url.pathname)
      .reply(statusCode, responseBody);
  };

  let setMocksForServices = (
    services: DataTypes.IService[],
    responseBody: any,
  ) => {
    services.forEach((service: DataTypes.IService) => {
      let url: URL = new URL(service.uri);
      setupServiceHealthcheckMock(url, 200, responseBody);
    });
  };

  test('broadcasts message if one of the services has changed state', done => {
    setupServicesAndConnectionsMocks();

    let services: DataTypes.IService[] = getFileContentAsJSON(
      './mocks/services.json',
    );

    setMocksForServices(services, {status: 'Healthy'});

    let firstService: DataTypes.IService = services[0];

    server
      .start(CONFIG_URI)
      .then(result => {
        const wsc1: WebSocketClient = new WebSocketClient();
        wsc1.onMessage = data => {
          let msg: IMessage = JSON.parse(data);
          if (msg.reply == 'UPDATE') {
            const content: DataTypes.IServiceLastKnownState = msg.content as DataTypes.IServiceLastKnownState;
            expect(content.status).toBe(DataTypes.Status.UNHEALTHY);
            expect(server.getServicesStatusById(firstService.id)).toEqual(
              content,
            );
            done();
          }
        };
        wsc1.connect(connectionString).then(result => {
          nock.cleanAll();
          setMocksForServices(services, {status: 'Unhealthy'});
        });
      })
      .catch(error => {
        fail(error);
        done();
      });
  });

  test('marks service as unhealthy if healthcheck request fails', done => {
    setupServicesAndConnectionsMocks();

    let services: DataTypes.IService[] = getFileContentAsJSON(
      './mocks/services.json',
    );

    let firstService: DataTypes.IService = services[0];

    server
      .start(CONFIG_URI)
      .then(result => {
        const wsc1: WebSocketClient = new WebSocketClient();
        wsc1.onMessage = data => {
          let msg: IMessage = JSON.parse(data);
          if (msg.reply == 'UPDATE') {
            const content: DataTypes.IServiceLastKnownState = msg.content as DataTypes.IServiceLastKnownState;
            expect(content.status).toBe(DataTypes.Status.UNHEALTHY);
            expect(server.getServicesStatusById(firstService.id)).toEqual(
              content,
            );
            done();
          }
        };
        wsc1.connect(connectionString);
      })
      .catch(error => {
        fail(error);
        done();
      });
  });

  test('returns a readonly copy of the services statuses', () => {
    let services: DataTypes.IService[] = getFileContentAsJSON(
      './mocks/services.json',
    );
    let servicesStatus = server.getServicesStatus();
    expect(servicesStatus).not.toBe(server.getServicesStatus());
  });

  test('returns a readonly copy of a service status', () => {
    let services: DataTypes.IService[] = getFileContentAsJSON(
      './mocks/services.json',
    );
    let firstService: DataTypes.IService = services[0];
    let servicesStatus = server.getServicesStatusById(firstService.id);
    expect(servicesStatus).not.toBe(firstService);
  });

  const issueReloadRequest = (done: any) => {
    axios.post(`${LOCAL_SERVER}/reload`).catch(error => {
      fail(error);
      done();
    });
  };

  test('reloads services and connections config on demand', done => {
    setupServicesAndConnectionsMocks();
    server
      .start(CONFIG_URI)
      .then(result => {
        const wsc1: WebSocketClient = new WebSocketClient();
        wsc1.onMessage = data => {
          let msg: IMessage = JSON.parse(data);
          if (msg.reply == 'RELOADED') {
            done();
          }
        };
        wsc1.connect(connectionString).then(result => {
          setupServicesAndConnectionsMocks(
            200,
            200,
            'reloadServices.json',
            'reloadConnections.json',
          );
          issueReloadRequest(done);
        });
      })
      .catch(error => {
        fail(error);
        done();
      });
  });

  test('forwards config post requests to config server', done => {
    setupServicesAndConnectionsMocks();
    setupMockConfigAPI_POST('/config', 204, {});
    server.start(CONFIG_URI).then(result => {
      axios
        .post(`${LOCAL_SERVER}/config`)
        .catch(error => {
          fail(error);
        })
        .finally(() => {
          done();
        });
    });
  });
});
