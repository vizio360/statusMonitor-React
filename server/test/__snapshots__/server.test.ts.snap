// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status Monitoring Server handles not existing commands 1`] = `
Object {
  "reply": "COMMAND NOT RECOGNIZED",
}
`;

exports[`Status Monitoring Server sends all services status 1`] = `
Object {
  "content": Array [
    Object {
      "responseBody": "NOT YET VERIFIED",
      "serviceId": "1",
      "status": 0,
    },
    Object {
      "responseBody": "NOT YET VERIFIED",
      "serviceId": "2",
      "status": 0,
    },
    Object {
      "responseBody": "NOT YET VERIFIED",
      "serviceId": "3",
      "status": 0,
    },
  ],
  "reply": "CURRENT_STATES",
}
`;

exports[`Status Monitoring Server sends connections config 1`] = `
Object {
  "content": Array [
    Object {
      "sourceId": "3",
      "targetId": "5",
    },
    Object {
      "sourceId": "5",
      "targetId": "4",
    },
    Object {
      "sourceId": "6",
      "targetId": "4",
    },
    Object {
      "sourceId": "2",
      "targetId": "4",
    },
    Object {
      "sourceId": "7",
      "targetId": "4",
    },
    Object {
      "sourceId": "7",
      "targetId": "2",
    },
  ],
  "reply": "CONNECTIONS",
}
`;

exports[`Status Monitoring Server sends services config 1`] = `
Object {
  "content": Array [
    Object {
      "categories": Array [],
      "id": "1",
      "matcher": Object {
        "status": "Healthy",
      },
      "name": "serviceTest",
      "timeout": 0.5,
      "type": "",
      "uiProps": Object {
        "left": 7,
        "top": 68,
      },
      "uri": "http://service1/healthcheck",
    },
    Object {
      "categories": Array [
        "cascdasdc",
      ],
      "id": "2",
      "matcher": Object {
        "status": "Healthy",
      },
      "name": "CRM",
      "timeout": 20,
      "type": "CRM",
      "uiProps": Object {
        "left": 712,
        "top": 441,
      },
      "uri": "http://service2/healthcheck",
    },
    Object {
      "categories": Array [],
      "id": "3",
      "matcher": Object {
        "status": "Healthy",
      },
      "name": "DataHub",
      "timeout": 20,
      "type": "API",
      "uiProps": Object {
        "left": 218,
        "top": 75,
      },
      "uri": "http://service3/healthcheck",
    },
  ],
  "reply": "SERVICES",
}
`;
